Optimize how much data needs to be `chown`/`chmod`ed on container startup

pR0Ps opened the pull request:
This works around an issue where `chmod`/`chown` operations inside a container can be extremely slow when using the `overlay2` storage driver, resulting in 10min+ container startup times.

It modifies the owner of the webapp when building the container so that only the `data` and `extensions` directories (which are commonly mapped as volumes into the container) have to be modified by the `access-permissions.sh` script at container startup.

When not running via docker the behaviour of the `access-permissions.sh` script is unchanged.

This reduced the container startup time for me from ~15 minutes to a few seconds.

Closes #7764

Changes proposed in this pull request:

- Uses the `COPY` command in the dockerfiles to set the group of the copied data as `www-data`
- Changes the docker entrypoint script to call `access-permissions.sh --only-userdirs` 
- Changes `access-permissions.sh` to only `chown`/`chmod` the `data` and `extensions` directories instead of the entire webapp if the `--only-userdirs` flag is provided.

How to test the feature manually:

1. Run the container and ensure it starts up quickly and properly.

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [ ] unit tests written (optional if too hard)
- [ ] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/edge/docs/en/developers/04_Pull_requests.md).

pR0Ps committed:
Optimize how much data needs to be `chown`/`chmod`ed on container startup

This works around an issue where `chmod`/`chown` operations inside a
container can be extremely slow when using the `overlay2` storage
driver, resulting in 10min+ container startup times.

It modifies the owner of the webapp when building the container so that
only the `data` and `extensions` directories (which are commonly mapped
as volumes into the container) have to be modified by the
`access-permissions.sh` script at container startup.

When not running via docker the behaviour of the `access-permissions.sh`
script is unchanged.

Alkarex commented on cli/access-permissions.sh:
```diff
@@ -11,11 +11,18 @@ if [ "$(id -u)" -ne 0 ]; then
 
+# If specified, only modify the data and extension dirs
+if [ "${1:-}" = "--only-userdirs" ]; then
+	to_update="./data ./extensions"
```
It would be nice to include the `DATA_PATH` environment variable as target, if defined:
https://github.com/FreshRSS/FreshRSS/blob/9faf2c1fa3c8c3ed82ff4c95733a018d0f9977a7/constants.php#L63-L69

Edit: relevant for the cases both with `--only-userdirs` and without

pR0Ps replied:
Added processing of `DATA_PATH` environment variable both with and without `--only-userdirs`.

Alkarex commented:
From CI:

```
Running shellcheck on cli/access-permissions.sh

In cli/access-permissions.sh line 22:
chown -R :www-data $to_update
                   ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
chown -R :www-data "$to_update"


In cli/access-permissions.sh line 25:
chmod -R g+rX $to_update
              ^--------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean: 
chmod -R g+rX "$to_update"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

pR0Ps committed:
Take DATA_PATH environment variable into account when fixing permissions

pR0Ps committed:
Revert change to using bash for arrays

(the alpine image doesn't include `bash`)

Alkarex committed:
A few more improvements

Alkarex commented on Docker/Dockerfile:
```diff
@@ -16,7 +16,7 @@ RUN apt-get update && \
 WORKDIR /var/www/FreshRSS
 
-COPY . /var/www/FreshRSS
+COPY --chown=:www-data . /var/www/FreshRSS
```
For some reason, this syntax fails on my Docker `Docker version 27.5.1, build 27.5.1-0ubuntu3` with the following error message:
> unable to convert uid/gid chown string to host mapping: can't find uid for user : no such user:

pR0Ps replied:
Strange, I checked the docs and there's no mention of the user-less syntax that I used but it seemed to work. Your version is definitely more correct though.

Alkarex commented:
I have made a [few more improvements](https://github.com/FreshRSS/FreshRSS/pull/7793/commits/ac75ce17d53ef8fc01e173b6cb4bb3016f57b945). Please double-check

pR0Ps committed:
Slightly tweak reapply permissions variable

- lowercase to indicate it's not an env variable
- use 0/1 to address potentially-irrational paranoia about unset variables

pR0Ps commented:
Your changes look good. I just pushed a very minor change to the `REAPPLY_PERMISSIONS` variable - feel free to discard it if you don't agree.

The `edge` branch is broken for me right now (I think it's OIDC-related, ~~haven't tracked it down fully yet~~ **EDIT:** See #7798) but when I apply all the commits in the PR on top of the current `latest` branch, it works perfectly.

pR0Ps committed:
Remove conditional logic to skip reapplying permissions

Also documents why in a comment so it's not missed in the future.

pR0Ps commented:
Actually, I found an issue with the latest commit while trying it with a fresh install. 

This line: https://github.com/FreshRSS/FreshRSS/blob/1f8273803f632b40c260721cb3257ae4e2251f5a/Docker/entrypoint.sh#L50
will copy some data in the `./data` directory as root and will therefore always need another call to `./cli/access-permissions.sh --only-userdirs` to fix the permissions.

pR0Ps commented:
I've pushed a fix for the above issue (removing the conditional reapplication of permissions).

In order to test this latest iteration of the PR fully, I ran FreshRSS in a very basic configuration (no OIDC) with both a fresh install (empty data directory) and an install with existing data:
```yaml
services:
  freshrss:
    build:
      context: https://github.com/pR0Ps/FreshRSS.git#edge
      dockerfile: Docker/Dockerfile-Newest
    container_name: freshrss
    restart: unless-stopped
    volumes:
      - /data/freshrss/data:/var/www/FreshRSS/data
      - /data/freshrss/extensions:/var/www/FreshRSS/extensions
    ports:
      - 8080:80
    environment:
      CRON_MIN: 8,38
      OIDC_ENABLED: 0
```

I also tested all combinations of this PR based on the `edge`/`latest` branch , as well as specifying `Dockerfile-Newest`/`Dockerfile` - all of them worked.

Alkarex commented:
Thanks again! Please add a line for you in https://github.com/FreshRSS/FreshRSS/blob/edge/CREDITS.md
